h5pt.go

```diff
@@ -82,170 +97,143 @@ func (t *Table) ReadPackets(start, nrecords int, data interface{}) error {
-		offset := C.size_t(rt.Field(i).Offset)
-
-		switch ft {
-		case reflect.Int8:
-			val := C.int8_t(int8(f.Int()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Uint8:
-			val := C.uint8_t(uint8(f.Uint()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Int16:
-			val := C.uint8_t(int16(f.Int()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Uint16:
-			val := C.uint8_t(uint16(f.Uint()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Int32:
-			val := C.uint8_t(int32(f.Int()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Uint32:
-			val := C.uint8_t(uint32(f.Uint()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Int64:
-			val := C.int64_t(int64(f.Int()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Uint64:
-			val := C.uint64_t(uint64(f.Uint()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Float32:
-			val := C.float(float32(f.Float()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Float64:
-			val := C.double(float64(f.Float()))
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		case reflect.Bool:
-			val := C.uchar(0)
-			if f.Bool() {
-				val = 1
-			}
-			dataPtr = unsafe.Pointer(&val)
-			dataSize = C.size_t(unsafe.Sizeof(val))
-		default:
-			err = fmt.Errorf("hdf5: Could not append struct member %s "+
-				"to PacketTable", ft)
-			return nil, err
-		}
-
-		// If the earlier allocated memory is not enough, we will increase it
-		// by another 64 bytes.
-		if memSize < (dataSize + offset) {
-			memSize += 64
-			ptr = C.realloc(ptr, memSize)
-			C.memset(unsafe.Pointer(uintptr(ptr)+uintptr(offset)), 0, 64)
-		}
-		C.memcpy(unsafe.Pointer(uintptr(ptr)+uintptr(offset)), dataPtr, dataSize)
-	}
-
-	return ptr, nil
-}
-
-// Append appends packets to the end of a packet table.
-func (t *Table) Append(data interface{}) (err error) {
+// Extracts the values that have to be appended to a PT. Writes them to mem, which
+// is then written to the HDF5 file.
+func extractValues(ad *appendData, data interface{}) (err error) {
 	rv := reflect.Indirect(reflect.ValueOf(data))
-	rp := reflect.Indirect(reflect.ValueOf(&data))
 	rt := rv.Type()
-	cNrecords := C.size_t(1)
-	cData := unsafe.Pointer(nil)
+
+	var dataPtr unsafe.Pointer
+	var dataSize C.size_t
 
 	switch rt.Kind() {
 	case reflect.Slice, reflect.Array:
 		for i := 0; i < rv.Len(); i++ {
-			if err = t.Append(rv.Index(i).Interface()); err != nil {
-				return err
+			if err = extractValues(ad, rv.Index(i).Interface()); err != nil {
+				return
 			}
 		}
-		return nil
+		ad.numRecords = C.size_t(rv.Len())
+		return
```

Comment:
Please don't use naked returns like this; make this and others like it, `return nil` if nil, otherwise `return err`.